src/analysis/convex/strict_convex_space.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Yury Kudryashov
+-/
+import analysis.convex.strict
+import analysis.convex.topology
+import analysis.normed_space.ordered
+import analysis.normed_space.pointwise
+
+/-!
+# Strictly convex spaces
+
+This file defines strictly convex spaces. A normed space is strictly convex if all closed balls are
+strictly convex. This does **not** mean that the norm is strictly convex (in fact, it never is).
+
+## Main definitions
+
+`strict_convex_space`: a typeclass saying that a given normed space over a normed linear ordered
+field (e.g., `ℝ` or `ℚ`) is strictly convex. The definition requires strict convexity of a closed
+ball of positive radius with center at the origin; strict convexity of any other closed ball follows
+from this assumption.
+
+## Main results
+
+In a strictly convex space, we prove
+
+- `strict_convex_closed_ball`: a closed ball is strictly convex.
+- `combo_mem_ball_of_ne`, `open_segment_subset_ball_of_ne`, `norm_combo_lt_of_ne`:
+  a nontrivial convex combination of two points in a closed ball belong to the corresponding open
+  ball;
+- `norm_add_lt_of_not_same_ray`, `same_ray_iff_norm_add`, `dist_add_dist_eq_iff`:
+  the triangle inequality `dist x y + dist y z ≤ dist x z` is a strict inequality unless `y` belongs
+  to the segment `[x -[ℝ] z]`.
+
+We also provide several lemmas that can be used as alternative constructors for `strict_convex ℝ E`:
+
+- `strict_convex_space.of_strict_convex_closed_unit_ball`: if `closed_ball (0 : E) 1` is strictly
+  convex, then `E` is a strictly convex space;
+
+- `strict_convex_space.of_norm_add`: if `∥x + y∥ = ∥x∥ + ∥y∥` implies `same_ray ℝ x y` for all
+  `x y : E`, then `E` is a strictly convex space.
+
+## Implementation notes
+
+While the definition is formulated for any normed linear ordered field, most of the lemmas are
+formulated only for the case `𝕜 = ℝ`.
+
+## Tags
+
+convex, strictly convex
+-/
+
+open set metric
+open_locale convex pointwise
+
+/-- A *strictly convex space* is a normed space where the closed balls are strictly convex. We only
+require balls of positive radius with center at the origin to be strictly convex in the definition,
+then prove that any closed ball is strictly convex in `strict_convex_closed_ball` below.
+
+See also `strict_convex_space.of_strict_convex_closed_unit_ball`. -/
+class strict_convex_space (𝕜 E : Type*) [normed_linear_ordered_field 𝕜] [normed_group E]
+  [normed_space 𝕜 E] : Prop :=
+(strict_convex_closed_ball : ∀ r : ℝ, 0 < r → strict_convex 𝕜 (closed_ball (0 : E) r))
```

Comment:
I feel like this is a weird middle ground to have as the definition. What I mean is this: strict convexity follows from strict convexity of the unit ball, and in a strictly convex space, all balls are strictly convex. So, in the definition, I would expect one of those two (unit ball or all balls), and presumably the former with a theorem for the latter. As it is now, whenever someone makes a `strict_convex_space` they will always opt for the constructor lemma `strict_convex_space.of_strict_convex_closed_unit_ball`, but instead you could just make that the definition. So, to be clear, I'm asking for:
```lean
class strict_convex_space (𝕜 E : Type*) [normed_linear_ordered_field 𝕜] [normed_group E]
  [normed_space 𝕜 E] : Prop :=
(strict_convex_closed_ball : strict_convex 𝕜 (closed_ball (0 : E) (1 : ℝ))
```
With a new proof of `strict_convex_closed_ball` and deleting `strict_convex_space.of_strict_convex_closed_unit_ball`. Of course, if you implement this, ignore the documentation comments above about "all balls"

Reply:
Is the point that strict convexity of the unit ball only implies strict convexity of all other balls if `E` is an `ℝ`-vector space? If so, is there an easy counterexample? In that case, just say so and ignore this comment.

Reply:
For me, the main question is "do we want to have `𝕜` in this file?"
All useful theorems only work for `𝕜 = ℝ`. See https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2311794.20Strictly.20convex.20spaces

Reply:
Proving strict convexity of a ball at the origin is significantly easier than proving strict convexity of the unit ball, because you can use the norm in place of the distance to `0`. Proving strict convexity of the unit ball does not seem significantly easier than proving strict convexity of any ball at the origin.

I believe `𝕜 = ℚ` also works, because the main restriction to "unit ball → any ball" is that your points might reach only a discrete set of norms.  Hence I believe some polynomial ring over some space will make "unit ball → any ball" fail.